### PR TITLE
Add hover label and hovered bar highlighting

### PR DESCRIPTION
**Pull request by akopich**
Tests: Manual
Bug: http://b/224507208

### CODE REVIEW

**Comment by karupayun** [comment]
screenshot?

File: src/OrbitQt/HistogramWidget.cpp [hunk 1]
```diff
@@ -146,33 +152,65 @@ static void DrawVerticalAxis(QPainter& painter, const QPoint& axes_intersection,
...
+                          int vertical_axis_length, double max_freq, uint64_t min_value,
+                          const std::optional<int>& histogram_hover_x) {
   std::vector<int> widths =
       orbit_qt::GenerateHistogramBinWidths(histogram.counts.size(), horizontal_axis_length);
 
-  int left_x = axes_intersection.x() + kLineWidth / 2 +
-               ValueToAxisLocation(histogram.min, horizontal_axis_length, min_value, histogram.max);
   int color_index = 0;
+  std::optional<orbit_qt::BarData> hovered_bar_data;
+  int left = axes_intersection.x() + kLineWidth / 2 +
```
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:178]: Why renaming from left_x to left?

File: src/OrbitQt/HistogramWidget.cpp [hunk 2]
```diff
@@ -146,33 +152,65 @@ static void DrawVerticalAxis(QPainter& painter, const QPoint& axes_intersection,
...
+             ValueToAxisLocation(histogram.min, horizontal_axis_length, min_value, histogram.max);
 
   // If the number of bins exceeds the width of histogram in pixels, `widths[i]` might be zero.
   // In such case we plot the bar on top of the previous one
+  // Because of that we kepp track of hovered_bar_data (multiple bars may be hovered at once),
```
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:183]: Keep.

File: src/OrbitQt/HistogramWidget.cpp [hunk 3]
```diff
@@ -146,33 +152,65 @@ static void DrawVerticalAxis(QPainter& painter, const QPoint& axes_intersection,
...
+  // as we render the tallest bar, the hover label shows the highest frequency
```
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:184]: Nit: Consider adding a full stop:
"... hovered at once).
As we render ... "

File: src/OrbitQt/HistogramWidget.cpp [hunk 4]
```diff
@@ -146,33 +152,65 @@ static void DrawVerticalAxis(QPainter& painter, const QPoint& axes_intersection,
...
   for (size_t i = 0; i < histogram.counts.size(); ++i) {
     double freq = GetFreq(histogram, i);
     if (freq > 0) {
-      const QPoint top_left(left_x,
-                            axes_intersection.y() - kLineWidth -
-                                ValueToAxisLocation(freq, vertical_axis_length, 0, max_freq));
-      const QPoint lower_right(left_x + std::max(widths[i] - 1, 0),
-                               axes_intersection.y() - kLineWidth);
+      const int top = axes_intersection.y() - kLineWidth -
```
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:188]: top_y

File: src/OrbitQt/HistogramWidget.cpp [hunk 5]
```diff
@@ -146,33 +152,65 @@ static void DrawVerticalAxis(QPainter& painter, const QPoint& axes_intersection,
...
+                      ValueToAxisLocation(freq, vertical_axis_length, 0, max_freq);
+      const int right = left + std::max(widths[i] - 1, 0);
```
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:190]: right_x

File: src/OrbitQt/HistogramWidget.cpp [hunk 6]
```diff
@@ -146,33 +152,65 @@ static void DrawVerticalAxis(QPainter& painter, const QPoint& axes_intersection,
...
+      const QPoint top_left(left, top);
+      const QPoint lower_right(right, axes_intersection.y() - kLineWidth);
```
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:192]: bottom_right

File: src/OrbitQt/HistogramWidget.cpp [hunk 7]
```diff
@@ -279,7 +317,7 @@ void HistogramWidget::paintEvent(QPaintEvent* /*event*/) {
 
   const orbit_statistics::Histogram& histogram = histogram_stack_.top();
 
-  QPoint axes_intersection(kLeftMargin, Height() - kBottomMargin);
+  const QPoint axes_intersection(kLeftMargin, Height() - kBottomMargin);
```
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.cpp:320]: Try to make unrelated cleaning changes in a different PR to facilitate review process.

File: src/OrbitQt/HistogramWidget.h [hunk 8]
```diff
@@ -22,6 +22,11 @@
 
 namespace orbit_qt {
 
+struct BarData {
+  double frequency;
+  int top;  // the y-coordinate for the top of the bar to be rendered
```
Comment (karupayun) [inline src/OrbitQt/HistogramWidget.h:27]: Nit: I would rename to top_y_pos and erase the comment.

**Review by karupayun — COMMENTED** [review]
Everything looks nice, just minor comments while waiting for a screenshot.

**Comment by akopich** [comment]
screenshot/4sx23NehN8VKA2R

File: src/OrbitQt/HistogramWidget.cpp (hunk 1 above)
Comment (akopich) [reply src/OrbitQt/HistogramWidget.cpp:178]: Because `_x` is redundant. It's clear left/right correspond to x-axis and top/bottom correspond to y-axis.

File: src/OrbitQt/HistogramWidget.h (hunk 8 above)
Comment (akopich) [reply src/OrbitQt/HistogramWidget.h:27]: Maybe, `top_pos` would be enough (see my comment above)?

File: src/OrbitQt/HistogramWidget.cpp (hunk 1 above)
Comment (karupayun) [reply src/OrbitQt/HistogramWidget.cpp:178]: I would rather prefer to keep them. "Clear" is a strong word.

**Comment by karupayun** [comment]
> http://screenshot/4sx23NehN8VKA2R

Nice, I like it! Please consider bolding the text as in the prototype.

**Comment by akopich** [comment]
This time it's bold.

http://screenshot/69WdMReWpexdePy

**Review by beckerhe — APPROVED** [review]
LGTM

**Review by karupayun — APPROVED** [review]
LGTM! Thanks for addressing the comments.
